### DISCUSSION TITLE
posts/Go/functional_options_pattern/

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#341] [Announcements]
https://www.liwenzhou.com/posts/Go/functional_options_pattern/ 

 李文周的Blog go 源码 设计模式 函数式 go-micro grpc 选项模式 Go语言

**Comment by wanghaolong613** [comment]
这个模式6啊

**Comment by waxtears** [comment]
为什么不用set方法，newOption().SetA("沙河娜扎").SetB(B).SetC(250)

**Comment by yudidi** [comment]
> @waxtears
> 为什么不用set方法，newOption().SetA("沙河娜扎").SetB(B).SetC(250)

有道理...

**Comment by Q1mi** [comment]
> @waxtears
> 为什么不用set方法，newOption().SetA("沙河娜扎").SetB(B).SetC(250)

1. 如果只想在初始化的时候设置相关字段的话，提供了SetXx之类的方法就可以随时修改字段值。
2. 如果对外提供的是接口类型而不是直接暴露该结构体的场景下，你就需要为接口类型添加很多的SetXx方法。

**Comment by depost** [comment]
nice,,,,

**Comment by depost** [comment]
源码里面少了一行，type Option func(o *Options)

**Comment by doulei123** [comment]
我也感觉和setXX很像啊。。

**Comment by asdlin** [comment]
秒的

**Comment by mingo-liu** [comment]
巧妙！
